tool/planet/agent.go

```diff
@@ -250,23 +249,46 @@ func updateDNS(conf *LeaderConfig, hostname string, newMasterIP string) error {
 	return nil
 }
 
-var electedUnits = []string{
+var controlPlaneUnits = []string{
 	"kube-controller-manager.service",
 	"kube-scheduler.service",
 	"kube-apiserver.service",
 }
 
-func unitsCommand(command string) error {
-	log.Debugf("executing %v on %v", command, electedUnits)
+func startUnits(ctx context.Context) error {
+	log.Debug("Start control plane units.")
 	var errors []error
-	for _, unit := range electedUnits {
-		cmd := exec.Command("/bin/systemctl", command, unit)
-		log.Debugf("executing %v", cmd)
-		out, err := cmd.CombinedOutput()
+	for _, unit := range controlPlaneUnits {
+		logger := log.WithField("unit", unit)
+		err := systemctlCmd(ctx, "start", unit)
 		if err != nil {
 			errors = append(errors, err)
 			// Instead of failing immediately, complete start of other units
-			log.Warningf("failed to execute %v: %s\n%v", cmd, out, trace.DebugReport(err))
+			logger.WithError(err).Warn("Failed to start unit.")
+		}
+	}
+	return trace.NewAggregate(errors...)
+}
+
+func stopUnits(ctx context.Context) error {
+	log.Debug("Stop control plane units.")
+	var errors []error
+	for _, unit := range controlPlaneUnits {
+		logger := log.WithField("unit", unit)
+		err := systemctlCmd(ctx, "stop", unit)
```

Comment:
Thinking about this for a moment, I wonder with the stop commands blocking, whether it could take a long time for the apiserver to even begin shutting down here, blocked on the controller and scheduler shutdowns. 

I'm also curious whether this could be causing us any sanity issues, since it's possible for the startup on a new node after election change can race the shutdown of the old node (we tell the processes there is only a single instance running, although I suspect locks are taken regardless). It might make sense to move this to an errgroup, with each stop running it's own goroutine. On the other hand I'm not sure I have enough faith in systemd to accept rapid fire commands like that.

Reply:
I don't think running the commands in parallel is a good idea unless we can retry the stop for apiserver since it's in the dependency list for both scheduler and controller manager services or just ensure they're both stopped before stopping the apiserver which the old code is doing anyways.
One thing this is not doing is ensuring that the services really did shut down - if they had not, i.e. systemctl came back with an error from dbus (we've seen issues with the startup), the services might not have shut down. I'm not sure how much to invest into that as opposed to [having the reconciler loop](https://github.com/gravitational/planet/pull/616) which should fix that as well. I'd be considering this as a precursor to the reconciler loop that applies the relevant patch in the short term.
Regarding the lock between the new and the old node - yes, I suppose this could race and end up unexpectedly which I believe could still be fixed with a proper reconciler loop.